### PR TITLE
Instrument 2.0.1 - Instrument 2 with ComponentInfo

### PR DESCRIPTION
**Pull request by OwenArnold**
The features here have been developed against a long-running branch `instrument_2_0_1`. They have been incrementally tested reviewed and merged in the form of feature branches into `instrument_2_0_1`. The branch is now ready for merging into master.

Fixes:

#19278 
#20002 
#20163

Main features addressed here are:

* Expand `ComponentInfo` for storing rotations and positions
* Replaced Mantid's component rotation and translation functionality to use `ComponentInfo`
* Simplified `Detector` and `Component` regarding parameter map usage
* Eliminated `DetectorInfo`'s compatibility modes around `Instrument 1.0` for setting positions and rotations for non-detectors and navigating the instrument tree.
* Used `InstrumentVisitor` mechanism to parse `Instrument 1.0` for both `ComponentInfo` and `DetectorInfo`
* Optimised Performance using `SCDCalibratePanels` as a known benchmark.

Authored by Owen Arnold and @SimonHeybrock 

Fixes a number of features that have been developed against the `instrument_2_0_1` branch.

**Release Notes** 

Most changes internal. We will modify the functional/performance section of the release notes when we have a better indication of performance improvements across the whole framework.

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####
As an internal change, unit system-tests provide a good indication that this is functionally correct

##### Performance Tests #####
Please pay attention to performance tests against current master. Any facility specific edge-cases should be tested, especially where load performance, and read-write performance is known to be critical.

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
@reviewers: Note that we have internally reviewed all changes before merging into our feature branch, i.e., this PR is a second review and more general test.

@OwenArnold Maybe you can link the original PRs in the PR text for reference?

File: Framework/Beamline/inc/MantidBeamline/ComponentInfo.h
Comment (martyngigg) [inline Framework/Beamline/inc/MantidBeamline/ComponentInfo.h]: Is this comment still true? If this is now in Geometry then it can't reference ExperimentInfo at all can it?

**Review by martyngigg — APPROVED** [review]
I had a quick glance but as this as already had reviews on it I think it's good to go.

File: Framework/Beamline/inc/MantidBeamline/ComponentInfo.h
Comment (OwenArnold) [inline Framework/Beamline/inc/MantidBeamline/ComponentInfo.h]: The wrappers `Geometry::DetectorInfo` in this case, are now the actual owners. I'll update comment.

Comment (martyngigg) [inline Framework/Beamline/inc/MantidBeamline/ComponentInfo.h]: :+1: Thanks.

**Review by VickieLynch — APPROVED** [review]
I compared single crystal reduction with TOPAZ between mantid38 and this branch:
Load took the same time
FilterBadPulses was a factor of 4.7 slower
LoadIsawDetCal was a factor of 4 faster
ConvertToMD was a factor of 2.2 slower
BinMD was 36% slower

**Pull request merged**
